### ISSUE TITLE
F* wrongly allows non-linear patterns

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#1121]
Recently discovered with @kyoDralliam that F* allows non-linear patterns:
```
let foo() : Tot int =
  match (1,2) with
  | (x,x) -> x
```
This is hugely unintuitive for users. Is this match complete? Does it equate the two elements of the pair? If not is this returning 1 or 2?

**Comment by cpitclaudel** [comment]
FTR "yes", "no", 2.  I agree it's unintuitive.

Ocaml and F# complain:
```
# let foo : int =
  match (1,2) with
  | (x,x) -> x;;

    Characters 42-43:
    | (x,x) -> x;;
         ^
Error: Variable x is bound several times in this matching
```

```
> let foo : int =
  match (1,2) with
  | (x,x) -> x;;

    | (x,x) -> x;;
  -------^

stdin(3,8): error FS0038: 'x' is bound twice in this pattern
```

**Comment by nikswamy** [comment]
Probably an easy fix in `FStar.ToSyntax.ToSyntax.fs`, in `desugar_data_pat`. There's a `check_linear_pattern_variables` at the end of that function, which apparently does not do the right thing.

**Comment by mtzguido** [comment]
A PR for this: #1309

**Comment by mtzguido** [comment]
I merged it, fixed by 2554cc1cd62c75c56be99a21ce850b365640633d
